### PR TITLE
Redesign provider interfaces

### PR DESCRIPTION
**Pull request by apmasell**
This redesigns the provider classes to use the Jackson dark magic to load
consumable resources, provisioners, and workflow engines directly using
Jackson.

The providers now export the classes to use with the matching `"type"` string
name. Jackson handles the deserialization int the correct subclass.

### CODE REVIEW

File: vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/ConsumableResource.java [hunk 1]
```diff
@@ -46,4 +90,11 @@
    */
   ConsumableResourceResponse request(
       String workflowName, String workflowVersion, String vidarrId, Optional<JsonNode> input);
+
+  /**
+   * Called to initialise this workflow engine.
```
Comment (callunity) [inline vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/ConsumableResource.java:95]: Won't most of these be something other than workflow engines?

**Review by callunity — APPROVED** [review]
This *is* some dark magic. I like it.

File: vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/ConsumableResource.java (hunk 1 above)
Comment (apmasell) [reply vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/ConsumableResource.java:95]: Yes. I did some copy pasta.

**Review by avarsava — APPROVED** [review]

**Comment by callunity** [comment]
FWIW I'm happy to merge this and update my code, but @avarsava may be more affected than me by this.

**Comment by avarsava** [comment]
> FWIW I'm happy to merge this and update my code, but @avarsava may be more affected than me by this.

I can update my code if this change is needed

File: vidarr-niassa/src/main/java/ca/on/oicr/gsi/vidarr/niassa/NiassaWorkflowEngine.java [hunk 2]
```diff
@@ -304,6 +311,37 @@ public JsonNode run(
     return null;
   }
 
+  @Override
+  public void startup() {
+    // Set up Postgres config if it has not been configured already
```
Comment (avarsava) [inline vidarr-niassa/src/main/java/ca/on/oicr/gsi/vidarr/niassa/NiassaWorkflowEngine.java:316]: This says 'if it has not been configured already' but the check for pgDataSource being null is gone

Comment (apmasell) [reply vidarr-niassa/src/main/java/ca/on/oicr/gsi/vidarr/niassa/NiassaWorkflowEngine.java]: Fixed the comment. `startup` will only get run once.

**Review by avarsava — APPROVED** [review]

**Pull request merged**